artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/Message.java

```diff
@@ -463,30 +463,40 @@ default void reencode() {
    }
 
    default void referenceOriginalMessage(final Message original, String originalQueue) {
-      String queueOnMessage = original.getAnnotationString(Message.HDR_ORIGINAL_QUEUE);
+      String originalQueueOnMessage = original.getAnnotationString(Message.HDR_ORIGINAL_QUEUE);
 
-      if (queueOnMessage != null) {
-         setAnnotation(Message.HDR_ORIGINAL_QUEUE, queueOnMessage);
-      } else if (originalQueue != null) {
-         setAnnotation(Message.HDR_ORIGINAL_QUEUE, originalQueue);
+      if (originalQueueOnMessage != null) {
+         setBreadcrumb(Message.HDR_ORIGINAL_QUEUE, originalQueueOnMessage);
       }
 
+      setAnnotation(Message.HDR_ORIGINAL_QUEUE, originalQueue);
+
       Object originalID = original.getAnnotation(Message.HDR_ORIG_MESSAGE_ID);
 
       if (originalID != null) {
```

Comment:
The code suggests, that setBreadcrumb might happen for Message.HDR_ORIGINAL_QUEUE but not for Message.HDR_ORIGINAL_ADDRESS and Message.HDR_ORIG_MESSAGE_ID if for example originalID is null.
As mentioned in past, I am no expert in Artemis therefore I cannot judge if such situation would ever happen, however by analysing this code only, there is a chance that breadcrumbs will run out of sync, as they are set under two separate conditions.
Will that be a problem however?

Reply:
Messages are moved between queues its only needed to track queues

Reply:
It seems theoretically possible, but I can't currently conceive of an actual use-case where this would happen. If you can write a test that demonstrates this happening then I'll change it.